profiling-ffi/src/exporter.rs

```diff
@@ -165,128 +169,125 @@ impl From<RequestBuildResult> for Result<Box<Request>, String> {
+            RequestBuildResult::Err(err) => Err(err.to_string()),
         }
     }
 }
 
 /// If successful, builds a `ddog_prof_Exporter_Request` object based on the
 /// profile data supplied. If unsuccessful, it returns an error message.
 ///
-/// The main use of the `ddog_prof_Exporter_Request` object is to be used in
-/// `ddog_prof_Exporter_send`, which will take ownership of the object. Do not
-/// pass it to `ddog_prof_Exporter_Request_drop` after that!
-///
 /// # Safety
 /// The `exporter`, `additional_stats`, and `endpoint_stats` args should be
 /// valid objects created by this module, except NULL is allowed.
 #[no_mangle]
 #[must_use]
 pub unsafe extern "C" fn ddog_prof_Exporter_Request_build(
-    exporter: Option<NonNull<ProfileExporter>>,
+    exporter: *mut ProfileExporter,
     start: Timespec,
     end: Timespec,
     files: Slice<File>,
-    additional_tags: Option<&ddcommon_ffi::Vec<Tag>>,
-    endpoints_stats: Option<&profiled_endpoints::ProfiledEndpointsStats>,
+    additional_tags: *const ddcommon_ffi::Vec<Tag>,
+    endpoints_stats: *const profiled_endpoints::ProfiledEndpointsStats,
     timeout_ms: u64,
 ) -> RequestBuildResult {
-    match exporter {
-        None => RequestBuildResult::Err(anyhow::anyhow!("no exporter was provided").into()),
+    match exporter.as_ref() {
+        None => RequestBuildResult::Err(anyhow::anyhow!("exporter was null").into()),
         Some(exporter) => {
             let timeout = std::time::Duration::from_millis(timeout_ms);
             let converted_files = into_vec_files(files);
-            let tags = additional_tags.map(|tags| tags.iter().map(Tag::clone).collect());
+            let tags = additional_tags
+                .as_ref()
+                .map(|tags| tags.iter().map(Tag::clone).collect());
 
-            match exporter.as_ref().build(
+            match exporter.build(
                 start.into(),
                 end.into(),
                 converted_files.as_slice(),
                 tags.as_ref(),
-                endpoints_stats,
+                endpoints_stats.as_ref(),
                 timeout,
             ) {
-                Ok(request) => RequestBuildResult::Ok(Box::into_raw(Box::new(request))),
+                Ok(request) => {
+                    RequestBuildResult::Ok(NonNull::new_unchecked(Box::into_raw(Box::new(request))))
+                }
                 Err(err) => RequestBuildResult::Err(err.into()),
             }
         }
     }
 }
 
 /// # Safety
-/// The `request` must be valid, or null. Notably, it is invalid if it's been
-/// sent to `ddog_prof_Exporter_send` or `ddog_prof_Exporter_Request_drop`
-/// already.
+/// The `request` must be valid, or null. After this call, the `request` will
+/// be invalid.
 #[no_mangle]
 pub unsafe extern "C" fn ddog_prof_Exporter_Request_drop(request: *mut Request) {
     if !request.is_null() {
-        drop(Box::from_raw(request));
+        drop_in_place(request)
     }
 }
```

Comment:
So one thing I don't understand in the change to `ddog_prof_Exporter_send` to receive a `request: *mut *mut Request` is -- why is `send` different than `drop`? E.g. why have `*mut *mut` in one but `*mut` in the other?

Reply:
I'll think more about this, thanks.

Reply:
Pushed this up.